### PR TITLE
MODINVSTOR-441 Preceding titles ability to add "unconnected" and "connected" title (create, edit)

### PR DESCRIPTION
**Pull request by Svitlana-Zmiivska**
https://issues.folio.org/browse/MODINVSTOR-441

Create new table for preceding/succeeding titles
Create new crud endpoints for preceding/succeeding titles

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -1343,6 +1370,31 @@
...
+    {
+      "permissionName": "inventory-storage.preceding-succeeding-titles.item.put",
+      "displayName": "inventory storage - modify preceding-succeeding-title",
+      "description": "modify preceding-succeeding-title in storage"
+    },
+    {
+      "permissionName": "inventory-storage.preceding-succeeding-titles.item.delete",
+      "displayName": "inventory storage - delete individual preceding-succeeding-title",
+      "description": "delete individual preceding-succeeding-title from storage"
+    },
```
Comment (bohdan-suprun) [inline descriptors/ModuleDescriptor-template.json]: We should add those permissions to `inventory-storage.all` as well.

File: ramls/instanceprecedingsucceedingtitle.json [hunk 2]
```diff
@@ -0,0 +1,63 @@
+{
```
Comment (bohdan-suprun) [inline ramls/instanceprecedingsucceedingtitle.json]: Is it possible to name file with `-` separator, e.g. `instance-preceding-succeeding-title.json` or similar?

File: ramls/instanceprecedingsucceedingtitles.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      "description": "List of instance relationships",
+      "id": "precedingSucceedingTitles",
+      "type": "array",
+      "items": {
+        "type": "object",
+        "$ref": "instanceprecedingsucceedingtitle.json"
+      }
+    },
+    "totalRecords": {
+      "type": "integer"
```
Comment (bohdan-suprun) [inline ramls/instanceprecedingsucceedingtitles.json]: Please add description.

File: src/main/java/org/folio/rest/impl/PrecedingSucceedingTitleAPI.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+import javax.ws.rs.core.Response;
+import java.util.Map;
+
+public class PrecedingSucceedingTitleAPI implements org.folio.rest.jaxrs.resource.PrecedingSucceedingTitles {
+
+  private static final String PRECEDING_SUCCEEDING_TITLE_TABLE = "preceding_succeeding_title";
+
+  @Override
+  public void getPrecedingSucceedingTitles(int offset, int limit, String query, String lang,
```
Comment (bohdan-suprun) [inline src/main/java/org/folio/rest/impl/PrecedingSucceedingTitleAPI.java]: Please add `@Validate` annotation.

File: src/main/java/org/folio/rest/impl/PrecedingSucceedingTitleAPI.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    PrecedingSucceedingTitle entity, Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    if (precedingOrSucceedingInstanceIdIsEmpty(entity)) {
+      handleError(asyncResultHandler);
+      return;
+    }
+    PgUtil.put(PRECEDING_SUCCEEDING_TITLE_TABLE, entity, precedingSucceedingTitleId, okapiHeaders, vertxContext,
+      PutPrecedingSucceedingTitlesByPrecedingSucceedingTitleIdResponse.class, asyncResultHandler);
+  }
+
+  private boolean precedingOrSucceedingInstanceIdIsEmpty(PrecedingSucceedingTitle entity) {
```
Comment (bohdan-suprun) [inline src/main/java/org/folio/rest/impl/PrecedingSucceedingTitleAPI.java]:
```suggestion
  private boolean precedingAndSucceedingInstanceEmpty(PrecedingSucceedingTitle entity) {
```

File: src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+
+    PrecedingSucceedingTitle precedingSucceedingTitle = new PrecedingSucceedingTitle(
+      instance1Id, instance2Id, TITLE, HRID, identifiers);
+
+    Response response = createReferenceRecord(
+      precedingSucceedingTitleUrl("").getPath(), precedingSucceedingTitle);
+    assertThat(response.getStatusCode(), is(HttpURLConnection.HTTP_CREATED));
+
+    assertPrecedingSucceedingTitleIsCreated(response, instance1Id, instance2Id, TITLE, HRID, identifiers);
+    deletePrecedingSucceedingTitle(response.getJson().getString("id"));
```
Comment (bohdan-suprun) [inline src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java]: You can use `org.folio.rest.api.StorageTestSuite#deleteAll(tenantId, tableName)` to remove entire table in the `@BeforeClass` method instead.

File: src/main/java/org/folio/rest/impl/PrecedingSucceedingTitleAPI.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  public void getPrecedingSucceedingTitles(int offset, int limit, String query, String lang,
+    Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    PgUtil.get(PRECEDING_SUCCEEDING_TITLE_TABLE, PrecedingSucceedingTitle.class, PrecedingSucceedingTitles.class, query, offset, limit,
+      okapiHeaders, vertxContext, GetPrecedingSucceedingTitlesResponse.class, asyncResultHandler);
+  }
+
+  @Override
+  public void postPrecedingSucceedingTitles(String lang, PrecedingSucceedingTitle entity,
+    Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    if (precedingOrSucceedingInstanceIdIsEmpty(entity)) {
```
Comment (bohdan-suprun) [inline src/main/java/org/folio/rest/impl/PrecedingSucceedingTitleAPI.java]: Is it possible to add a constraint for this to Postgre as well?

File: src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+    JsonObject instance1Response = createInstance("Title One", INSTANCE_TYPE_ID_TEXT);
+    JsonObject instance2Response = createInstance("Title Two", INSTANCE_TYPE_ID_TEXT);
+    String instance1Id = instance1Response.getString(Instance.ID_KEY);
+    String instance2Id = instance2Response.getString(Instance.ID_KEY);
+
+    JsonArray identifiers = new JsonArray();
+    identifiers.add(identifier(UUID_ISBN, "9781473619777"));
+
+    PrecedingSucceedingTitle precedingSucceedingTitle = new PrecedingSucceedingTitle(
```
Comment (bohdan-suprun) [inline src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java]: Are we going to store title, HRID and identifiers for connected case as well? What should happen if a succeeding/preceding instance is got updated with some new identifiers?

Comment (Svitlana-Zmiivska) [reply src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java]: No. we aren't. Removed these fields from test

File: ramls/instanceprecedingsucceedingtitle.json (hunk 2 above)
Comment (Svitlana-Zmiivska) [reply ramls/instanceprecedingsucceedingtitle.json]: Yes, it is. In my opinion it is more readable also. However, all other json files use this style. And I am not sure it is good idea to use another style only for preceding-succeeding-titles json file.

Comment (bohdan-suprun) [reply ramls/instanceprecedingsucceedingtitle.json]: Well, yes. Let's keep naming in the same way for now. 

(I think we should have a tech debt to fix all the names).

File: src/main/java/org/folio/rest/impl/PrecedingSucceedingTitleAPI.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    }
+    PgUtil.put(PRECEDING_SUCCEEDING_TITLE_TABLE, entity, precedingSucceedingTitleId, okapiHeaders, vertxContext,
+      PutPrecedingSucceedingTitlesByPrecedingSucceedingTitleIdResponse.class, asyncResultHandler);
+  }
+
+  private boolean precedingAndSucceedingInstanceEmpty(PrecedingSucceedingTitle entity) {
+    return ObjectUtils.isEmpty(entity.getPrecedingInstanceId()) && ObjectUtils.isEmpty(entity.getSucceedingInstanceId());
+  }
+
+  private void handleError(Handler<AsyncResult<Response>> asyncResultHandler) {
```
Comment (bohdan-suprun) [inline src/main/java/org/folio/rest/impl/PrecedingSucceedingTitleAPI.java]:
```suggestion
  private void handlePrecedingAndSucceedingEmptyError(Handler<AsyncResult<Response>> asyncResultHandler) {
```

File: src/main/resources/templates/db_scripts/alterPrecedingSucceedingTitle.sql [hunk 10]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE ${myuniversity}_${mymodule}.preceding_succeeding_title
```
Comment (bohdan-suprun) [inline src/main/resources/templates/db_scripts/alterPrecedingSucceedingTitle.sql]: I think we need a drop constraint statement right before creation. Because subsequent execution of the script will fail with a 'constraint already exists' error.

File: src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+import org.folio.rest.support.Response;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+public class PrecedingSucceedingTitleTest extends ResourceTestBase {
```
Comment (bohdan-suprun) [inline src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java]: Could you please make sure that following edge case is covered as well:
* PUT/DELETE/GET fails with error when specified wrong id;

File: src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+      instance1Response.getString(Instance.ID_KEY), instance2Response.getString(Instance.ID_KEY),
+      TITLE, HRID, identifiers);
+    Response postResponse = createReferenceRecord(entityPath, precedingSucceedingTitle);
+    assertThat(postResponse.getStatusCode(), is(HttpURLConnection.HTTP_CREATED));
+
+    String entityUUID = postResponse.getJson().getString("id");
+
+    String updateProperty = PrecedingSucceedingTitle.TITLE_KEY;
+
+    testGetPutDeletePost(entityPath, entityUUID, precedingSucceedingTitle, updateProperty);
```
Comment (bohdan-suprun) [inline src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java]: Hi @Svitlana-Zmiivska, I'm not sure that the `preceding-succeeding` API falls under ReferenceRecordAPI term.

Usually `ReferenceRecordAPI` means a very basic CRUD API that used as a storage for static data (e.g. instance relationship types, identifiers types, locations, etc.).

Comment (Svitlana-Zmiivska) [reply src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java]: changed

File: src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+    JsonObject instance1Response = createInstance("Title One", INSTANCE_TYPE_ID_TEXT);
+    JsonObject instance2Response = createInstance("Title Two", INSTANCE_TYPE_ID_TEXT);
+    String instance1Id = instance1Response.getString(Instance.ID_KEY);
+    String instance2Id = instance2Response.getString(Instance.ID_KEY);
+
+    PrecedingSucceedingTitle precedingSucceedingTitle = new PrecedingSucceedingTitle(
+      instance1Id, instance2Id, null, null, null);
+
+    Response response = createReferenceRecord(
+      precedingSucceedingTitleUrl("").getPath(), precedingSucceedingTitle);
```
Comment (bohdan-suprun) [inline src/test/java/org/folio/rest/api/PrecedingSucceedingTitleTest.java]: Does it make sense to create a client for precedingSucceeding API? similar that we have for items/holdings/etc.
